nomos-services/data-availability/dispersal/src/adapters/network/libp2p.rs

```diff
@@ -100,4 +136,89 @@ where
                     as BoxStream<'static, Result<(BlobId, Self::SubnetworkId), DynError>>
             })
     }
+
+    async fn get_blob_samples(
+        &self,
+        blob_id: BlobId,
+        subnets: &[SubnetworkId],
+        cooldown: Duration,
+    ) -> Result<(), DynError> {
+        let expected_count = subnets.len();
+        let mut success_count = 0;
+
+        let mut pending_subnets: HashSet<SubnetworkId> = subnets.iter().copied().collect();
+
+        let (stream_sender, stream_receiver) = oneshot::channel();
+
+        self.outbound_relay
+            .send(DaNetworkMsg::Subscribe {
+                kind: DaNetworkEventKind::Sampling,
+                sender: stream_sender,
+            })
+            .await
+            .map_err(|(error, _)| error)?;
+
+        self.start_sampling(blob_id, subnets).await?;
+
+        let stream = match stream_receiver.await {
+            Ok(stream) => stream,
+            Err(error) => return Err(Box::new(error) as DynError),
+        };
+
+        enum SampleOutcome {
+            Success(u16),
+            Retry(u16),
+        }
+
+        let mut stream = tokio_stream::StreamExt::filter_map(stream, move |event| match event {
+            DaNetworkEvent::Sampling(SamplingEvent::SamplingSuccess {
+                blob_id: event_blob_id,
+                ref blob,
+            }) if event_blob_id == blob_id => Some(SampleOutcome::Success(blob.column_idx)),
+            DaNetworkEvent::Sampling(SamplingEvent::SamplingError { ref error }) => match error {
+                SamplingError::Protocol {
+                    error,
+                    subnetwork_id,
+                    ..
+                } if error.blob_id == blob_id => Some(SampleOutcome::Retry(*subnetwork_id)),
+                SamplingError::Deserialize {
+                    blob_id: event_blob_id,
+                    subnetwork_id,
+                    ..
+                } if *event_blob_id == blob_id => Some(SampleOutcome::Retry(*subnetwork_id)),
+                SamplingError::BlobNotFound {
+                    blob_id: event_blob_id,
+                    subnetwork_id,
+                    ..
+                } if *event_blob_id == blob_id => Some(SampleOutcome::Retry(*subnetwork_id)),
+                _ => None,
+            },
+            _ => None,
+        });
```

Comment:
I think it would be clearer if you pre-filter the sample `blob_id`. Maybe a method can be added to `SamplingEvent` to retrieve the `blob_id` if all variants have it now.

Reply:
https://github.com/logos-co/nomos-node/pull/1104/commits/bf69e927c0cb4520f163ffa3a6315497663cacc5